python/pyarrow/dataset.py

```diff
@@ -232,168 +219,406 @@ def _ensure_format(obj):
+    )
+
+    if isinstance(fs_or_uri, str):
+        # instantiate the file system from an uri, if the uri has a path
+        # component then it will be treated as a path prefix
+        filesystem, prefix = FileSystem.from_uri(fs_or_uri)
+        is_local = isinstance(filesystem, LocalFileSystem)
+        prefix = _normalize_path(filesystem, prefix)
+        if prefix:
+            # validate that the prefix is pointing to a directory
+            prefix_info = filesystem.get_file_info([prefix])[0]
+            if prefix_info.type != FileType.Directory:
+                raise ValueError(
+                    "The path component of the filesystem URI must point to a "
+                    "directory but it has a type: `{}`. The path component "
+                    "is `{}` and the given filesystem URI is `{}`".format(
+                        prefix_info.type.name, prefix_info.path, fs_or_uri
+                    )
+                )
+            filesystem = SubTreeFileSystem(prefix, filesystem)
+        return filesystem, is_local
+    elif isinstance(fs_or_uri, (LocalFileSystem, _MockFileSystem)):
+        return fs_or_uri, True
+    elif isinstance(fs_or_uri, FileSystem):
+        return fs_or_uri, False
+    else:
+        raise TypeError(
+            '`filesystem` argument must be a FileSystem instance or a valid '
+            'file system URI'
+        )
+
+
+def _ensure_multiple_sources(paths, filesystem=None):
     """
-    Create a factory which can be used to build a Dataset.
+    Treat a list of paths as files belonging to a single file system
+
+    If the file system is local then also validates that all paths
+    are referencing existing *files* otherwise any non-file paths will be
+    silently skipped (for example on a remote filesystem).
 
     Parameters
     ----------
-    path_or_paths : str, pathlib.Path, or list of those
-        Path to a file or to a directory containing the data files, or
-        a list of paths.
-    filesystem : FileSystem, default None
-        By default will be inferred from the path.
-    partitioning : Partitioning or PartitioningFactory or str or list of str
-        The partitioning scheme specified with the ``partitioning()``
-        function. A flavor string can be used as shortcut, and with a list of
-        field names a DirectionaryPartitioning will be inferred.
-    format : str, default None
-        Currently only "parquet" is supported.
+    paths : list of path-like
+        Note that URIs are not allowed.
+    filesystem : FileSystem or str, optional
+        If an URI is passed, then its path component will act as a prefix for
+        the file paths.
 
     Returns
     -------
-    FileSystemDatasetFactory
+    (FileSystem, list of str)
+        File system object and a list of normalized paths.
+
+    Raises
+    ------
+    TypeError
+        If the passed filesystem has wrong type.
+    FileNotFoundError
```

Comment:
You don't need to detail all this. This is pointless cognitive overhead for the user, IMHO. Just document `IOError`.